File: ticket-tracker-java/src/main/java/com/tickettracker/service/MasterDataService.java
package com.tickettracker.service;

import com.tickettracker.dao.MasterDataDAO;
import com.tickettracker.exception.ValidationException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.SQLException;
import java.util.List;

public class MasterDataService {

    private static final Logger logger = LoggerFactory.getLogger(MasterDataService.class);
    private final MasterDataDAO masterDataDAO;

    public MasterDataService() {
        this.masterDataDAO = new MasterDataDAO();
    }

    public List<MasterDataDAO.MapItem> getAll(String type) throws Exception {
        return getAll(type, null);
    }

    public List<MasterDataDAO.MapItem> getAll(String type, byte[] moduleId) throws Exception {
        String tableName = getTableName(type);
        return masterDataDAO.findAll(tableName, moduleId);
    }

    public MasterDataDAO.MapItem add(String type, String name) throws Exception {
        return add(type, name, null);
    }

    public MasterDataDAO.MapItem add(String type, String name, byte[] moduleId) throws Exception {
        if (name == null || name.trim().isEmpty()) {
            throw new ValidationException("Name is required");
        }
        String trimmed = name.trim();
        String tableName = getTableName(type);
        MasterDataDAO.MapItem existing = masterDataDAO.findByName(tableName, trimmed, moduleId);
        if (existing != null) {
            throw new ValidationException("\"" + trimmed + "\" already exists");
        }
        logger.info("Adding master {} item: {}", type, trimmed);
        return masterDataDAO.add(tableName, trimmed, moduleId);
    }

    public void remove(String type, byte[] id) throws Exception {
        String tableName = getTableName(type);
        masterDataDAO.delete(tableName, id);
        logger.info("Removed master {} item", type);
    }

    public void toggleActive(String type, byte[] id, boolean isActive) throws Exception {
        String tableName = getTableName(type);
        masterDataDAO.toggleActive(tableName, id, isActive);
    }

    public String getConfigValue(String key) throws Exception {
        return masterDataDAO.getConfigValue(key);
    }

    public void setConfigValue(String key, String value, String description) throws Exception {
        masterDataDAO.setConfigValue(key, value, description);
        logger.info("Updated config: {} = {}", key, value);
    }

    public String getModuleCode(byte[] moduleId) throws Exception {
        return masterDataDAO.getModuleCode(moduleId);
    }

    public void setModuleCode(byte[] moduleId, String moduleCode) throws Exception {
        masterDataDAO.setModuleCode(moduleId, moduleCode);
        logger.info("Updated module code for module");
    }

    public List<String> getActiveLocationNames() throws Exception {
        return masterDataDAO.findActiveNames("master_locations");
    }

    public List<String> getActivePropertyNames() throws Exception {
        return masterDataDAO.findActiveNames("master_properties");
    }

    public List<String> getActiveCategoryNames(byte[] moduleId) throws Exception {
        return masterDataDAO.findActiveNames("master_categories", moduleId);
    }

    public String generateTicketNumber(String locationPrefix, String moduleCode) throws Exception {
        String companyCode = masterDataDAO.getConfigValue("company_code");
        if (companyCode == null || companyCode.isEmpty()) {
            companyCode = "NMDC";
        }
        String loc3 = locationPrefix;
        if (loc3 == null || loc3.isEmpty()) {
            List<String> activeLocations = masterDataDAO.findActiveNames("master_locations");
            loc3 = activeLocations.isEmpty() ? "LOC" : activeLocations.get(0);
        }
        if (loc3.length() > 3) {
            loc3 = loc3.substring(0, 3);
        }
        loc3 = loc3.toUpperCase();

        int nextCounter = masterDataDAO.getNextCounter(loc3, moduleCode);
        String paddedCounter = String.format("%06d", nextCounter);

        String ticketNumber = String.format("TKT-%s%s-%s-%s", companyCode, loc3, moduleCode, paddedCounter);
        logger.info("Generated ticket number: {}", ticketNumber);
        return ticketNumber;
    }

    private String getTableName(String type) throws ValidationException {
        switch (type) {
            case "categories": return "master_categories";
            case "departments": return "master_departments";
            case "locations": return "master_locations";
            case "properties": return "master_properties";
            default:
                throw new ValidationException("Invalid master data type: " + type);
        }
    }
}

File: src/services/masterDataService.ts
import { supabase, isSupabaseAvailable } from '../lib/supabase';

export interface MasterItem {
  id: string;
  name: string;
  is_active: boolean;
  display_order: number;
  created_at: string;
  updated_at: string;
}

export interface MasterConfig {
  key: string;
  value: string;
  description: string;
}

export type MasterDataType = 'categories' | 'departments' | 'locations' | 'properties';

const TABLE_MAP: Record<MasterDataType, string> = {
  categories: 'master_categories',
  departments: 'master_departments',
  locations: 'master_locations',
  properties: 'master_properties',
};

const FALLBACK_CATEGORIES = [
  'Civil Maintenance',
  'Electrical Maintenance',
  'Plumbing & Sanitary',
  'Carpentry',
  'HVAC / Air Conditioning',
  'Water Supply',
  'Sewage & Drainage',
  'Road & External Area',
  'Housekeeping, Fire & Safety',
  'Security Systems',
  'Street Lighting',
  'Utility Services',
];

const FALLBACK_LOCATIONS = ['Location01', 'Location02'];
const FALLBACK_PROPERTIES = ['PROP001', 'PROP002'];

const FALLBACK_DEPARTMENTS: string[] = [];

export class MasterDataService {
  static async getAll(type: MasterDataType, moduleId?: string): Promise<MasterItem[]> {
    if (!isSupabaseAvailable()) {
      return this.getFallback(type);
    }
    try {
      const table = TABLE_MAP[type];
      let query = supabase.from(table).select('*');
      if (moduleId && type === 'categories') {
        query = query.eq('module_id', moduleId);
      }
      const { data, error } = await query.order('display_order', { ascending: true });

      if (error) throw error;
      return (data || []) as MasterItem[];
    } catch (error) {
      console.error(`Error fetching master ${type}:`, error);
      return this.getFallback(type);
    }
  }

  static async getActive(type: MasterDataType, moduleId?: string): Promise<string[]> {
    const items = await this.getAll(type, moduleId);
    return items.filter(i => i.is_active).map(i => i.name);
  }

  static async add(type: MasterDataType, name: string, moduleId?: string): Promise<MasterItem | null> {
    if (!isSupabaseAvailable()) return null;
    try {
      const table = TABLE_MAP[type];
      let existingQuery = supabase.from(table).select('id').ilike('name', name);
      if (moduleId && type === 'categories') {
        existingQuery = existingQuery.eq('module_id', moduleId);
      }
      const { data: existing } = await existingQuery.maybeSingle();

      if (existing) throw new Error(`"${name}" already exists`);

      let maxOrderQuery = supabase.from(table).select('display_order');
      if (moduleId && type === 'categories') {
        maxOrderQuery = maxOrderQuery.eq('module_id', moduleId);
      }
      const { data: maxOrder } = await maxOrderQuery.order('display_order', { ascending: false }).limit(1).maybeSingle();

      const nextOrder = (maxOrder?.display_order || 0) + 1;

      const insertRow: Record<string, unknown> = { name, is_active: true, display_order: nextOrder };
      if (moduleId && type === 'categories') {
        insertRow.module_id = moduleId;
      }
      const { data, error } = await supabase
        .from(table)
        .insert([insertRow])
        .select()
        .single();

      if (error) throw error;
      return data as MasterItem;
    } catch (error) {
      console.error(`Error adding master ${type}:`, error);
      throw error;
    }
  }

  static async remove(type: MasterDataType, id: string): Promise<void> {
    if (!isSupabaseAvailable()) return;
    try {
      const table = TABLE_MAP[type];
      const { error } = await supabase.from(table).delete().eq('id', id);
      if (error) throw error;
    } catch (error) {
      console.error(`Error removing master ${type}:`, error);
      throw error;
    }
  }

  static async toggleActive(type: MasterDataType, id: string, isActive: boolean): Promise<void> {
    if (!isSupabaseAvailable()) return;
    try {
      const table = TABLE_MAP[type];
      const { error } = await supabase.from(table).update({ is_active: isActive }).eq('id', id);
      if (error) throw error;
    } catch (error) {
      console.error(`Error toggling master ${type}:`, error);
      throw error;
    }
  }

  static async getConfig(key: string): Promise<string | null> {
    if (!isSupabaseAvailable()) return null;
    try {
      const { data, error } = await supabase
        .from('master_config')
        .select('value')
        .eq('key', key)
        .maybeSingle();

      if (error) throw error;
      return data?.value || null;
    } catch (error) {
      console.error(`Error fetching config ${key}:`, error);
      return null;
    }
  }

  static async setConfig(key: string, value: string, description?: string): Promise<void> {
    if (!isSupabaseAvailable()) return;
    try {
      const { error } = await supabase
        .from('master_config')
        .upsert({ key, value, description }, { onConflict: 'key' });
      if (error) throw error;
    } catch (error) {
      console.error(`Error setting config ${key}:`, error);
      throw error;
    }
  }

  static async getModuleCode(moduleId: string): Promise<string> {
    if (!isSupabaseAvailable()) return 'TKT';
    try {
      const { data, error } = await supabase
        .from('modules')
        .select('config')
        .eq('id', moduleId)
        .maybeSingle();

      if (error) throw error;
      return data?.config?.moduleCode || 'TKT';
    } catch (error) {
      console.error('Error fetching module code:', error);
      return 'TKT';
    }
  }

  static async setModuleCode(moduleId: string, moduleCode: string): Promise<void> {
    if (!isSupabaseAvailable()) return;
    try {
      const { data: module, error: fetchError } = await supabase
        .from('modules')
        .select('config')
        .eq('id', moduleId)
        .maybeSingle();

      if (fetchError) throw fetchError;
      if (!module) throw new Error('Module not found');

      const updatedConfig = { ...(module.config || {}), moduleCode };
      const { error } = await supabase
        .from('modules')
        .update({ config: updatedConfig })
        .eq('id', moduleId);

      if (error) throw error;
    } catch (error) {
      console.error('Error setting module code:', error);
      throw error;
    }
  }

  static async generateTicketNumber(locationPrefix: string, moduleCode: string): Promise<string> {
    if (!isSupabaseAvailable()) {
      return `TKT-${Date.now()}`;
    }
    try {
      const companyCode = await this.getConfig('company_code') || 'NMDC';

      let effectivePrefix = locationPrefix;
      if (!effectivePrefix) {
        const activeLocations = await this.getActive('locations');
        effectivePrefix = activeLocations.length > 0 ? activeLocations[0] : 'LOC';
      }
      const loc3 = effectivePrefix.substring(0, 3).toUpperCase();

      const { data: existing, error: selectError } = await supabase
        .from('ticket_number_counter')
        .select('id, counter')
        .eq('location_prefix', loc3)
        .eq('module_code', moduleCode)
        .maybeSingle();

      if (selectError) throw selectError;

      let nextCounter: number;

      if (existing) {
        nextCounter = existing.counter + 1;
        const { error: updateError } = await supabase
          .from('ticket_number_counter')
          .update({ counter: nextCounter })
          .eq('id', existing.id);

        if (updateError) throw updateError;
      } else {
        nextCounter = 1;
        const { error: insertError } = await supabase
          .from('ticket_number_counter')
          .insert([{ location_prefix: loc3, module_code: moduleCode, counter: nextCounter }]);

        if (insertError) throw insertError;
      }

      const paddedCounter = String(nextCounter).padStart(6, '0');
      return `TKT-${companyCode}${loc3}-${moduleCode}-${paddedCounter}`;
    } catch (error) {
      console.error('Error generating ticket number:', error);
      return `TKT-${Date.now()}`;
    }
  }

  private static getFallback(type: MasterDataType): MasterItem[] {
    let names: string[] = [];
    if (type === 'categories') names = FALLBACK_CATEGORIES;
    else if (type === 'locations') names = FALLBACK_LOCATIONS;
    else if (type === 'properties') names = FALLBACK_PROPERTIES;

    return names.map((name, i) => ({
      id: `fallback-${type}-${i}`,
      name,
      is_active: true,
      display_order: i + 1,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
    }));
  }
}
